File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> results;
        
        unordered_map<string, vector<string>> map;
        
        for(auto& str : strs)
        {
            string original = str;
            sort(str.begin(), str.end());
            map[str].push_back(original);
        }
        
        for(auto const& [k,v] : map)
        {
            results.push_back(v);
        }
        
        return results;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) 
    {
        HashMap<String, ArrayList<String>> hashmap = new HashMap<>();
        List<List<String>> result = new ArrayList<>();
        
        for(String str : strs)
        {
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);
            
            if(hashmap.containsKey(key))
            {
                hashmap.get(key).add(str);
            }
            else
            {
                hashmap.put(key, new ArrayList<>());
                hashmap.get(key).add(str);
            }
        }
        
        for(String key : hashmap.keySet())
        {
            result.add(hashmap.get(key));
        }
        
        return result;
    }
}